phoenix-core/src/main/java/org/apache/phoenix/hbase/index/write/IndexWriterUtils.java

```diff
@@ -70,13 +70,13 @@
    public static final String HTABLE_KEEP_ALIVE_KEY = "hbase.htable.threads.keepalivetime";
 
    public static final String INDEX_WRITER_RPC_RETRIES_NUMBER = "phoenix.index.writes.rpc.retries.number";
-   /**
-    * Based on the logic in HBase's AsyncProcess, a default of 11 retries with a pause of 100ms
-    * approximates 48 sec total retry time (factoring in backoffs).  The total time should be less
-    * than HBase's rpc timeout (default of 60 sec) or else the client will retry before receiving
-    * the response
-    */
-   public static final int DEFAULT_INDEX_WRITER_RPC_RETRIES_NUMBER = 11;
+    /**
+     * Retry server-server index write rpc only once, and let the client retry the data write
+     * instead to avoid typing up the handler
+     */
+   // note in HBase 2+, numTries = numRetries + 1
+   // in prior versions, numTries = numRetries
+   public static final int DEFAULT_INDEX_WRITER_RPC_RETRIES_NUMBER = 1;
```

Comment:
We need to conditionally leave it at 11 versus use 1 based on the client version. Old clients wouldn't have the retry logic, so we'd need to continue retrying on the server.

Reply:
Maybe the easiest way to do that is in Indexer.start() you could do a getTable call on SYSTEM.CATALOG and check the table timestamp - if it's been upgraded to 4.14, use 1, otherwise use 11. Though it'd be an RPC, we'd know that SYSTEM.CATALOG would be cached in our server-side cache so it'd be relatively cheap.

Reply:
A simpler solution would be to only set INDEX_WRITER_RPC_RETRIES_NUMBER if it's not set (Configuration.setIfNotSet()). That way an installation could set the config param to 1 until they know that clients have been upgraded at which point they could remove the setting.

Reply:
Current installations may not have any property configured and are just using the default of 11.  Post-upgrade server side, they would again be using the default, which has changed to 1.  So I think we would still need to check the client version?
Unless you're saying the onus should be on the operator to set configs based on the version being used, but that seems more confusing.

Reply:
I was thinking it'd be ok if the onus was on the operator. I think we're in the minority in having a mix of old/new clients with zero downtime. If you think it's worth checking the client version (or checking if upgraded yet) and can come up with a reasonable way of doing that, that's fine too. Not sure it's worth the effort, though. Here's some ideas:

* add client version to IndexMaintainer (which is protobuf-ed, so it'd be doable). Then use the "correct" IndexBuilder (or config?) based on the client version pulled out of the IndexMaintainer. If you have separate IndexBuilder instances, then you'd get duplicate thread pools to do the writing too which isn't ideal. Maybe within the parallel writer you could use the correct config?
* less dynamic, but check the timestamp of system.catalog to know what version you're at. Somewhat scary to have an RPC in the start path of the coprocessor, though, as if system.catalog RS can't be reached, then you're kind of screwed (but many you are already anyway).
* least dynamic - put onus on operator and change to setIfUnset.

Reply:
I don't think we need any code change if we're relying on the operator.  As it is, we check the config INDEX_WRITER_RPC_RETRIES_NUMBER, and only if it's not there, we use the DEFAULT_INDEX_WRITER_RPC_RETRIES_NUMBER, which is now 1.  So we assume the operator will have the property set to 11 until all clients are upgraded, at which point the remove the property and it uses the default of 1.

Reply:
Good point.